custom-instrumentation/instrument-datastore/index.js

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright 2024 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+
+const newrelic = require('newrelic')
+const { SimpleDatastore } = require('./simple-datastore')
+const fastify = require('fastify')({ logger: true })
+
+const datastore = new SimpleDatastore()
+const { PORT: port = 3000, HOST: host = '127.0.0.1' } = process.env
+
+// We will be using the fastify web framework to host the
+// datastore, but any host (Docker, web framework, etc.) can be used
+fastify.post('/query', async (request, reply) => {
+  // Execute a single query
+  const query = 'SELECT * FROM users'
+  const result = await datastore.execute(query)
+  return reply.send({ result })
+})
+
+fastify.post('/batch', async (request, reply) => {
+  // Execute a batch of queries.
+  const queries = ['SELECT * FROM orders', 'SELECT * FROM products']
+  const result = await datastore.batch(queries)
+  return reply.send({ result })
+})
+
+fastify.post('/shutdown', async (request, reply) => {
+  // Close the connection to the datastore
+  datastore.close()
+  // Close fastify
+  fastify.close()
+  return reply.send({ status: 'Connection closed' })
+})
+
+fastify.addHook('onClose', async (instance) => {
+  // Finally shutdown the agent so it properly flushes all data
+  setTimeout(() => {
+    newrelic.shutdown({ collectPendingData: true }, () => process.exit(0))
+  }, 2500)
+})
```

Comment:
I don't think we need the `setTimeout` in the `onClose` handler. We should be closing out the datastore in that handler, though. The `setImmediate` will queue the server shutdown to come _after_ the reply has been sent.

```suggestion
fastify.post('/shutdown', async (request, reply) => {
  // Close fastify
  setImmediate(() => fastify.close())
  return reply.send({ status: 'Connection closed' })
})

fastify.addHook('onClose', async (instance) => {
  // Close the connection to the datastore
  datastore.close()
  // Finally shutdown the agent so it properly flushes all data
  newrelic.shutdown({ collectPendingData: true }, () => process.exit(0))
})
```

Reply:
I tried a similar approach, but then shutdown (and thus datastore close) no longer appears in New Relic as an instrumented transaction